### PR TITLE
Fix type expression nullability to be the declared nullability, and c…

### PR DESCRIPTION
**Pull request by 333fred**
…ompensate for when passed an indentifier that is parented by a NullableSyntax.

Fixes https://github.com/dotnet/roslyn/issues/40925.

@jcouv, this fixes the bug you showed, but I'm not convinced it's the cause of the simplification issues. Further, I'm not convinced by the arguments you made in the bug: the IDE needs to understand that it should not be calling `GetTypeInfo` on the underlying syntax node, and should instead be calling it on the containing syntax. If calling it on the syntax underlying a `NullableTypeSyntax` is core to how the IDE simplifier works here, the design is busted.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 1]
```diff
@@ -2185,6 +2185,18 @@ internal protected virtual CSharpSyntaxNode GetBindableSyntaxNode(CSharpSyntaxNo
                 }
             }
 
+            switch (node.Kind())
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: >switch (node.Kind()) [](start = 12, length = 20)

It feels like this is not the right place for the adjustment. The nodes have semantic meaning without the parent.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -7193,7 +7193,7 @@ public override BoundNode VisitTypeExpression(BoundTypeExpression node)
                 VisitTypeExpression(node.BoundContainingTypeOpt);
             }
 
-            SetNotNullResult(node);
+            SetResult(node, node.TypeWithAnnotations.ToTypeWithState(), node.TypeWithAnnotations);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >SetResult(node, node.TypeWithAnnotations.ToTypeWithState(), node.TypeWithAnnotations); [](start = 12, length = 86)

Not all type references have corresponding BoundTypeExpression nodes. For example, those used in declarations don't. If the goal of adjusting this line is to adjust SemanticModel's behavior, it feels like not all scenarios are going to be handled.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs [hunk 3]
```diff
@@ -3815,5 +3815,40 @@ void verify(DefaultExpressionSyntax expr, PublicNullableAnnotation expectedAnnot
                 Assert.Equal(expectedState, info.FlowState);
             }
         }
+
+        [Fact]
+        public void GetTypeInfoOnNullableType()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: >GetTypeInfoOnNullableType [](start = 20, length = 25)

Please add tests for type references in declarations.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs [hunk 4]
```diff
@@ -3815,5 +3815,40 @@ void verify(DefaultExpressionSyntax expr, PublicNullableAnnotation expectedAnnot
...
+        {
+            var source = @"
+#nullable enable
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: Is result going to be different in disable context?

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 5]
```diff
@@ -2185,6 +2185,18 @@ internal protected virtual CSharpSyntaxNode GetBindableSyntaxNode(CSharpSyntaxNo
...
+            {
+                case SyntaxKind.AliasQualifiedName:
+                case SyntaxKind.QualifiedName:
+                case SyntaxKind.IdentifierName:
+                    if (node.Parent.IsKind(SyntaxKind.NullableType))
+                    {
+                        node = node.Parent;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: >node = node.Parent; [](start = 24, length = 19)

Do we really want to do this for "int" in ```int?```, for example?

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: >GetTypeInfoOnNullableType [](start = 20, length = 25)

Please add tests for nullable value types.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

**Comment by jcouv** [comment]
@333fred I'll try stacking my changes on top of your PR to confirm the issues were related. Thanks

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: I will be reverting this part of the PR after more discussion with Jason and Julien offline. This is just not something the IDE should expect to do and whatever answer they get back is what they get bakc.

---
In reply to: [366363378](https://github.com/dotnet/roslyn/pull/40950#discussion_r366363378) [](ancestors = 366363378)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs (hunk 3 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: As I've reverted the `NullableTypeSyntax`, I don't think this is relevant anymore.

---
In reply to: [366375766](https://github.com/dotnet/roslyn/pull/40950#discussion_r366375766) [](ancestors = 366375766)

Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: Added.

---
In reply to: [366368793](https://github.com/dotnet/roslyn/pull/40950#discussion_r366368793) [](ancestors = 366368793)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: > Is result going to be different in disable context?

Please add test with disabled context

---
In reply to: [366369803](https://github.com/dotnet/roslyn/pull/40950#discussion_r366369803) [](ancestors = 366369803)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs [hunk 6]
```diff
@@ -3815,5 +3815,56 @@ void verify(DefaultExpressionSyntax expr, PublicNullableAnnotation expectedAnnot
...
+        IEnumerable<C> M3() => throw null!;
+    }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+
+            var tree = comp.SyntaxTrees.Single();
+            var model = comp.GetSemanticModel(tree, ignoreAccessibility: false);
+            var identifiers = tree.GetRoot().DescendantNodes().OfType<VariableDeclarationSyntax>().ToList();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: >identifiers [](start = 16, length = 11)

Please assert there are only 12 items in the list

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs [hunk 7]
```diff
@@ -3815,5 +3815,56 @@ void verify(DefaultExpressionSyntax expr, PublicNullableAnnotation expectedAnnot
...
+
+            for (int i = 0; i < 6; i += 2)
+            {
+                Assert.Equal(PublicNullableAnnotation.NotAnnotated, model.GetTypeInfo(identifiers[i].Type).Type.NullableAnnotation);
+                Assert.Equal(PublicNullableAnnotation.Annotated, model.GetTypeInfo(identifiers[i + 1].Type).Type.NullableAnnotation);
+            }
+
+            var outVars = tree.GetRoot().DescendantNodes().OfType<DeclarationExpressionSyntax>().ToList();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: >outVars [](start = 16, length = 7)

Please assert amount of items in the list.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs [hunk 8]
```diff
@@ -3815,5 +3815,56 @@ void verify(DefaultExpressionSyntax expr, PublicNullableAnnotation expectedAnnot
...
+            Assert.Equal(PublicNullableAnnotation.NotAnnotated, model.GetTypeInfo(outVars[0].Type).Type.NullableAnnotation);
+            Assert.Equal(PublicNullableAnnotation.Annotated, model.GetTypeInfo(outVars[1].Type).Type.NullableAnnotation);
+
+            var foreaches = tree.GetRoot().DescendantNodes().OfType<ForEachStatementSyntax>().ToList();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: >foreaches [](start = 16, length = 9)

Please assert amount of items in the list.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Scenarios for types in member declarations will not go through this code path. How are they handled?

---
In reply to: [366367464](https://github.com/dotnet/roslyn/pull/40950#discussion_r366367464) [](ancestors = 366367464)

**Comment by AlekseyTs** [comment]
I am assuming this PR no longer fixes #40925. At least not in the way the issue is stated. Could you please provide details about the problem we trying to fix now? What behavior we used to have and what behavior we will have after the change?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 9]
```diff
@@ -7143,7 +7144,7 @@ public override void VisitForEachIterationVariables(BoundForEachStatement node)
                         }
 
                         // In non-error cases we'll only run this loop a single time. In error cases we'll set the nullability of the VariableType multiple times, but at least end up with something
-                        SetAnalyzedNullability(node.IterationVariableType, new VisitResult(result, destinationType), isLvalue: true);
+                        SetAnalyzedNullability(node.IterationVariableType, new VisitResult(isVar ? result : destinationType.ToTypeWithState(), destinationType), isLvalue: true);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >isVar ? result : destinationType.ToTypeWithState() [](start = 91, length = 50)

It is not clear to me what is going on here. It appears we are adjusting (by comparison to what we used to do) an RValue type, but only for a non-var case. Why is this needed and why only for non-var case?

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: > As I've reverted the NullableTypeSyntax, I don't think this is relevant anymore.

Are value types never go through the code paths that are modified in this PR? Are we not interested in what GetTypeInfor returns for them?

---
In reply to: [366588484](https://github.com/dotnet/roslyn/pull/40950#discussion_r366588484) [](ancestors = 366588484,366375766)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: > Added.

Sorry for not making it clear that I meant in declarations outside of executable code. I.e. in member declarations, etc. 

---
In reply to: [366588587](https://github.com/dotnet/roslyn/pull/40950#discussion_r366588587) [](ancestors = 366588587,366368793)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: I think type parameters for constructed types and methods in an executable code also are not backed by BoundTypeExpression nodes. Consider testing them too.

---
In reply to: [366616630](https://github.com/dotnet/roslyn/pull/40950#discussion_r366616630) [](ancestors = 366616630,366588587,366368793)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs (hunk 6 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: 6, but yes.

---
In reply to: [366606901](https://github.com/dotnet/roslyn/pull/40950#discussion_r366606901) [](ancestors = 366606901)

**Comment by 333fred** [comment]
This turned out to not be the cause of the bug. Closing this PR out: https://github.com/dotnet/roslyn/pull/41028
